rfd/0222-access-list-templates.md

```diff
@@ -0,0 +1,258 @@
+}
+```
+
+### Implementation
+
+#### Create
+
+A templated access list can only be created if access list fields `type: template` and `template_config` defines at least one resource field with its labels set. Resource principals will not be required.
+
+Once these fields are validated, Teleport will take the template specifications and create system roles and then assign them to member and owner.
+
+##### System roles for short-term template
+
+- access: a role that defines the access to resources - this role is not assigned directly to anyone
+- reviewer: a role assigned to owner grants that allow them to review requests to resources defined in `access` role
+- requester: a role assigned to member grants that allow them to search and request for resources defined in `access` role
+
+##### System roles for long-term template
+
+- access: a role that defines the access to resources and is assigned to member grants
+
+##### Naming system roles
+
+In order to ensure uniqueness and help identifying which roles belong to access lists, the naming convention takes the following format:
+
+`<purpose>-acl-role-<access list metadata name (UID)>`
+
+| Parts                             |                      Explanation                      |              Example Values |
+| :-------------------------------- | :---------------------------------------------------: | --------------------------: |
+| \<purpose\>                       |     short word that describes the purpose of role     | requester, reviewer, access |
+| acl-role                          |             stands for "access list role"             |                         n/a |
+| <access list metadata name (UID)> | helps identify which access list this role belongs to |                         n/a |
+
+Example names of system roles, if an access list with metadata.name is set to `abcd1234`
+
+- requester-acl-role-abcd1234
+- reviewer-acl-role-abcd1234
+- access-acl-role-abcd1234
+
+#### Update
+
+The `type: template`, member grant, and owner grant will not be modifiable.
+
+Minus the fields that are already modifiable, `template_config` field is partly modifiable. The field that defines access to resources can be modified. The `oneof type` will not be modifiable eg: if an access list template was originally `short-term`, a user cannot change the template to `long-term`.
+
+For both templates, since only the `access` part of the `template_config` can be modified, the system role related to `access` will be updated.
+
+##### Update quirk
+
+There is a quirk where the `access` definition set on an access list might not be in sync with the actual role resource because we don't prevent users from editing system roles with `tctl` (the web UI does not allow reading/updating system roles). If such a case happens, ultimately the role resource is the source of truth. Any updates made to `template_config` will overwrite any previous edits directly made to the system roles.
```

Comment:
I guess we should have a reconciler watching `access_list` resources with `type=template` and creating roles. So manual changes would be reverted after resync period. That will also open a possibility to also watch roles and revert such manual changes nearly instantly (depending on the cluster load) if needed.

Reply:
i need to look more into this. i think it'll be similar to the reconciler for the [access list ineligible status](https://github.com/gravitational/teleport.e/blob/fedcc26b1d3e2487e6b5c17b7a18ce22940644d3/lib/accesslist/reconciler_ineligible_status.go)